### PR TITLE
Write manifest multi upstream

### PR DESCRIPTION
**Pull request by dappnodedev**
After implementing https://github.com/dappnode/DAppNodeSDK/pull/416 we need to merge this in order to modify the manifest properly in the bump action. This PR depends on https://github.com/dappnode/DNP_DAPPMANAGER/pull/1936 to be merged and schemas and types be published after that. Also, those versions need to be bumped here.

The old way of handling multi-upstream repos like `"upstreamVersion": ["0.1.0", "0.1.2"]` has been deprecated. Any package that has several upstream repos must define their versions and args like:
```
upstream: {
  version: string;
  repo: string;
  arg: string;
}[]
```

### CODE REVIEW

File: src/commands/githubActions/bumpUpstream/github/fetchGithubUpstreamVersion.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { Github } from "../../../../providers/github/Github.js";
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/github/fetchGithubUpstreamVersion.ts:1]: clear and concise 👍

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 2]
```diff
@@ -57,35 +54,34 @@ async function gaBumpUpstreamHandler({
   use_fallback: useFallback,
 }: CliCommandOptions): Promise<void> {
 
-  const { upstreamSettings, manifestData: { manifest, format }, compose, gitSettings, ethProvider } = await getInitialSettings({ dir, userEthProvider, useFallback });
-  printSettings(upstreamSettings, gitSettings, manifest, compose, ethProvider);
+  const { upstreamSettings, manifestData: { manifest, format: manifestFormat }, compose, gitSettings, ethProvider } = await getInitialSettings({ dir, userEthProvider, useFallback });
+  if (!upstreamSettings) {
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:58]: looks better

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 3]
```diff
@@ -57,35 +54,34 @@ async function gaBumpUpstreamHandler({
...
+    console.log("There are no upstream repos/versions defined in the manifest");
+    return;
+  }
+
+  printSettings(upstreamSettings, gitSettings, ethProvider);
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:63]: do we want to print settings even if "All version are up-to-date" ?

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 4]
```diff
@@ -57,35 +54,34 @@ async function gaBumpUpstreamHandler({
...
     console.log(`Branch ${branchName} already exists`);
     return;
   }
 
-  const composeVersionsToUpdate = updateComposeVersions(dir, compose, upstreamRepoVersions);
-  if (!composeVersionsToUpdate) {
-    console.log("All versions are up-to-date");
-    return;
-  }
+  updateComposeVersions(dir, compose, upstreamSettings);
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:78]: looks better not to return anything in this function 👍

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 5]
```diff
@@ -99,110 +95,80 @@ async function gaBumpUpstreamHandler({
...
-  for (const [serviceName, service] of Object.entries(compose.services)) {
+  for (const upstreamItem of upstreamSettings) {
 
-    if (typeof service.build !== "object" || !service.build.args)
-      continue;
+    for (const [, service] of Object.entries(newCompose.services))
 
-    for (const [argName, currentVersion] of Object.entries(service.build.args)) {
-      const upstreamVersionInfo = upstreamRepoVersions[argName];
+      if (typeof service.build !== "string" && service.build?.args && upstreamItem.arg in service.build.args)
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:114]: It makese senst to me, although this condition look complex, could you add a comment right above giving an example of each and why is it necessary?

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 6]
```diff
@@ -99,110 +95,80 @@ async function gaBumpUpstreamHandler({
...
  * @param {string} dir - Directory for the Compose file.
  * @param {Compose} compose - Original Docker Compose configuration.
- * @param {UpstreamRepoMap} upstreamRepoVersions - Mapping of dependencies to their new versions.
- * @return {ComposeVersionsToUpdate | null} - Details of updated versions or null.
+ * @param {UpstreamSettings[]} upstreamSettings - New versions for the Compose services.
  */
-function updateComposeVersions(dir: string, compose: Compose, upstreamRepoVersions: UpstreamRepoMap): ComposeVersionsToUpdate | null {
-  const newCompose = JSON.parse(JSON.stringify(compose)); // Deep copy
-  const versionsToUpdate: ComposeVersionsToUpdate = {};
+function updateComposeVersions(dir: string, compose: Compose, upstreamSettings: UpstreamSettings[]): void {
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:107]: I would rename it to `updateComposeUpstreamVersions`

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 7]
```diff
@@ -99,110 +95,80 @@ async function gaBumpUpstreamHandler({
...
+ * and optionally fetches a new version for the manifest. The updated manifest is returned.
+ * @param {Object} options - The options for updating the manifest.
+ * @param {Manifest} options.manifest - The manifest object to update.
+ * @param {UpstreamSettings[]} options.upstreamSettings - The new versions for the manifest.
+ * @param {string} options.dir - Directory path where the manifest will be saved.
+ * @param {string} options.ethProvider - Ethereum provider URL.
+ * @returns {Promise<Manifest>} The updated manifest object.
  */
-async function updateManifest({
+async function updateManifestVersions({
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:132]: On first sight this function does the same as `updateComposeVersions`, bumping upstreamVersion, but it does not. It also bumps the pkg itselve version.

These are 2 different actions and was thinking if we should have separated functions for each action, with the corresponding proper naming: `updateManifestUpstreamVersions` and `updateManifestPkgVersion`. Realize that `updateManifestUpstreamVersions` would be sync as `updateComposeUpstreamVersions` and `updateManifestPkgVersion` would be async

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 7 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:134]: I agree

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 5 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:118]: Done

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 6 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts]: Done

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 7 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts]: Done

**Pull request merged**